File: signalpilot/gateway/gateway/db/models/dashboards.py
"""Published dashboards: the org-scoped team gallery with versions and refreshes.

A chat writes ``artifacts/<name>.dashboard.json``. Publishing copies the spec
and its dataset files into dashboard-owned object keys and creates a
published dashboard with version 1. Every refresh or restore adds an
immutable version; ``current_version_id`` names the live one. A refresh that
fails validation never replaces the live version.

Schema changes go through Alembic (0028 introduced these tables). Keep the
column definitions here byte-identical to the migration; the parity test
diffs the two.
"""

from __future__ import annotations

import uuid
from datetime import UTC, datetime

from sqlalchemy import (
    JSON,
    Boolean,
    ForeignKey,
    Index,
    Integer,
    String,
    Text,
    UniqueConstraint,
    func,
)
from sqlalchemy.orm import Mapped, mapped_column

from .base import GatewayBase, TZDateTime


def _utcnow() -> datetime:
    return datetime.now(UTC)


def new_dashboard_id() -> str:
    return f"dash_{uuid.uuid4().hex}"


def new_version_id() -> str:
    return f"dver_{uuid.uuid4().hex}"


def new_refresh_id() -> str:
    return f"dref_{uuid.uuid4().hex}"


class GatewayPublishedDashboard(GatewayBase):
    """One dashboard in the team gallery. Independent of its source chat."""

    __tablename__ = "gateway_published_dashboards"

    id: Mapped[str] = mapped_column(String, primary_key=True, default=new_dashboard_id)
    org_id: Mapped[str] = mapped_column(String, nullable=False)
    project_id: Mapped[str | None] = mapped_column(String)
    # ^[a-z0-9][a-z0-9-]{0,63}$, unique per org.
    slug: Mapped[str] = mapped_column(String(64), nullable=False)
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    description: Mapped[str | None] = mapped_column(Text)
    # private | org | link
    visibility: Mapped[str] = mapped_column(String(10), nullable=False, default="private", server_default="private")
    # Set when visibility == link. 32 url-safe characters, never hashed: the
    # token is the link and the owner can read it back.
    share_token: Mapped[str | None] = mapped_column(String(64))
    created_by_user_id: Mapped[str] = mapped_column(String, nullable=False)
    source_conversation_id: Mapped[str | None] = mapped_column(String)
    source_file_id: Mapped[str | None] = mapped_column(String)
    current_version_id: Mapped[str | None] = mapped_column(
        String,
        ForeignKey(
            "gateway_published_dashboard_versions.id",
            name="fk_gw_pubdash_current_version",
            ondelete="SET NULL",
            use_alter=True,
        ),
    )
    # null = off. Allowed: 15, 60, 120, 240, 480, 720, 1440.
    refresh_interval_minutes: Mapped[int | None] = mapped_column(Integer)
    # "HH:MM" 24h wall-clock in refresh_timezone.
    refresh_anchor_time: Mapped[str | None] = mapped_column(String(5))
    refresh_timezone: Mapped[str] = mapped_column(String(64), nullable=False, default="UTC", server_default="UTC")
    # sql | agent
    refresh_mode: Mapped[str] = mapped_column(String(10), nullable=False, default="sql", server_default="sql")
    notify_on_failure: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True, server_default="true")
    next_refresh_at: Mapped[datetime | None] = mapped_column(TZDateTime)
    last_refresh_at: Mapped[datetime | None] = mapped_column(TZDateTime)
    # succeeded | failed | skipped
    last_refresh_status: Mapped[str | None] = mapped_column(String(10))
    chart_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0, server_default="0")
    created_at: Mapped[datetime] = mapped_column(TZDateTime, nullable=False, default=_utcnow, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(TZDateTime, nullable=False, default=_utcnow, server_default=func.now())
    archived_at: Mapped[datetime | None] = mapped_column(TZDateTime)

    __table_args__ = (
        UniqueConstraint("org_id", "slug", name="uq_gw_pubdash_org_slug"),
        UniqueConstraint("share_token", name="uq_gw_pubdash_share_token"),
        Index("ix_gw_pubdash_org", "org_id"),
        Index("ix_gw_pubdash_due", "next_refresh_at"),
    )


class GatewayPublishedDashboardVersion(GatewayBase):
    """An immutable snapshot: one spec object plus one object per file dataset."""

    __tablename__ = "gateway_published_dashboard_versions"

    id: Mapped[str] = mapped_column(String, primary_key=True, default=new_version_id)
    dashboard_id: Mapped[str] = mapped_column(
        String,
        ForeignKey("gateway_published_dashboards.id", name="fk_gw_pubdash_version_dashboard", ondelete="CASCADE"),
        nullable=False,
    )
    version_no: Mapped[int] = mapped_column(Integer, nullable=False)
    spec_key: Mapped[str] = mapped_column(Text, nullable=False)
    # {dataset name: object key}. Inline-row datasets are absent: they stay in the spec.
    dataset_keys: Mapped[dict] = mapped_column(JSON, nullable=False)
    # {dataset name: {row_count, byte_size, filename}}
    dataset_meta: Mapped[dict] = mapped_column(JSON, nullable=False)
    chart_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0, server_default="0")
    # publish | refresh | restore
    produced_by: Mapped[str] = mapped_column(String(10), nullable=False)
    # User id (publish, restore) or refresh id (refresh).
    producer_ref: Mapped[str | None] = mapped_column(String)
    created_at: Mapped[datetime] = mapped_column(TZDateTime, nullable=False, default=_utcnow, server_default=func.now())

    __table_args__ = (UniqueConstraint("dashboard_id", "version_no", name="uq_gw_pubdash_version_no"),)


class GatewayDashboardRefresh(GatewayBase):
    """One refresh attempt, scheduled or manual, in sql or agent mode."""

    __tablename__ = "gateway_dashboard_refreshes"

    id: Mapped[str] = mapped_column(String, primary_key=True, default=new_refresh_id)
    dashboard_id: Mapped[str] = mapped_column(
        String,
        ForeignKey("gateway_published_dashboards.id", name="fk_gw_pubdash_refresh_dashboard", ondelete="CASCADE"),
        nullable=False,
    )
    org_id: Mapped[str] = mapped_column(String, nullable=False)
    # sql | agent
    mode: Mapped[str] = mapped_column(String(10), nullable=False)
    # schedule | manual
    trigger: Mapped[str] = mapped_column(String(10), nullable=False)
    # queued | running | succeeded | failed | skipped
    status: Mapped[str] = mapped_column(String(10), nullable=False, default="queued", server_default="queued")
    scheduled_for: Mapped[datetime | None] = mapped_column(TZDateTime)
    started_at: Mapped[datetime | None] = mapped_column(TZDateTime)
    finished_at: Mapped[datetime | None] = mapped_column(TZDateTime)
    version_id: Mapped[str | None] = mapped_column(String)
    # Chat run and conversation for agent mode.
    run_id: Mapped[str | None] = mapped_column(String)
    conversation_id: Mapped[str | None] = mapped_column(String)
    error: Mapped[str | None] = mapped_column(Text)
    # Per-dataset outcome and the notification record.
    detail: Mapped[dict | None] = mapped_column(JSON)
    created_at: Mapped[datetime] = mapped_column(TZDateTime, nullable=False, default=_utcnow, server_default=func.now())

    __table_args__ = (
        Index("ix_gw_pubdash_refresh_dashboard", "dashboard_id", "created_at"),
        Index("ix_gw_pubdash_refresh_org", "org_id"),
        Index("ix_gw_pubdash_refresh_status", "mode", "status"),
    )

File: signalpilot/web/lib/api/dashboards.ts
// Published dashboards: the team gallery, versions, refreshes, and the
// publish action from a chat artifact. The gateway is the source of truth.

import type { DashboardSpec, DatasetRows } from "~/dashboard-renderer";

import { GATEWAY_URL, getAuthHeaders, request } from "./client";

export type DashboardVisibility = "private" | "org";
export type DashboardRefreshMode = "sql" | "agent";
export type DashboardRefreshStatus =
  | "queued"
  | "running"
  | "finalizing"
  | "succeeded"
  | "failed"
  | "skipped";

/** Statuses of a refresh that is still in flight; mirrors the gateway's ACTIVE_REFRESH_STATUSES. */
const DASHBOARD_ACTIVE_REFRESH_STATUSES: readonly DashboardRefreshStatus[] = [
  "queued",
  "running",
  "finalizing",
];

export function isActiveRefreshStatus(status: DashboardRefreshStatus): boolean {
  return DASHBOARD_ACTIVE_REFRESH_STATUSES.includes(status);
}
export type DashboardVersionOrigin = "publish" | "refresh" | "restore";

/** Allowed refresh intervals in minutes; null means off. */
const DASHBOARD_REFRESH_INTERVALS = [15, 60, 120, 240, 480, 720, 1440] as const;
export type DashboardRefreshInterval = (typeof DASHBOARD_REFRESH_INTERVALS)[number];

export type DashboardRefreshSettings = {
  interval_minutes: DashboardRefreshInterval | null;
  /** "HH:MM" 24h wall-clock in `timezone`; the schedule aligns to it. */
  anchor_time: string | null;
  timezone: string;
  mode: DashboardRefreshMode;
};

export type PublishedDashboard = {
  id: string;
  slug: string;
  name: string;
  description: string | null;
  visibility: DashboardVisibility;
  project_id: string | null;
  created_by_user_id: string;
  created_by_label: string;
  source_conversation_id: string | null;
  source_file_id: string | null;
  current_version_id: string | null;
  current_version_no: number | null;
  chart_count: number;
  refresh: DashboardRefreshSettings;
  notify_on_failure: boolean;
  next_refresh_at: string | null;
  last_refresh_at: string | null;
  last_refresh_status: "succeeded" | "failed" | "skipped" | null;
  created_at: string;
  updated_at: string;
  archived_at: string | null;
  can_edit: boolean;
};

export type DashboardDatasetMeta = {
  row_count: number;
  byte_size: number;
  filename: string;
};

export type DashboardVersion = {
  id: string;
  version_no: number;
  produced_by: DashboardVersionOrigin;
  producer_ref: string | null;
  chart_count: number;
  dataset_meta: Record<string, DashboardDatasetMeta>;
  created_at: string;
};

export type DashboardRefresh = {
  id: string;
  mode: DashboardRefreshMode;
  trigger: "schedule" | "manual";
  status: DashboardRefreshStatus;
  scheduled_for: string | null;
  started_at: string | null;
  finished_at: string | null;
  version_id: string | null;
  run_id: string | null;
  conversation_id: string | null;
  error: string | null;
  detail: Record<string, unknown> | null;
  created_at: string;
};

export type DashboardDetail = {
  dashboard: PublishedDashboard;
  versions: DashboardVersion[];
  refreshes: DashboardRefresh[];
};

export type DashboardBundle = {
  dashboard: PublishedDashboard;
  version: DashboardVersion;
  spec: DashboardSpec;
  datasets: Record<string, DatasetRows>;
};

export type PublishDashboardRequest = {
  name: string;
  slug?: string;
  description?: string;
  visibility: DashboardVisibility;
  refresh: DashboardRefreshSettings;
  notify_on_failure?: boolean;
  /** Publish a new version of an existing dashboard instead of creating one. */
  target_dashboard_id?: string;
};

export type UpdateDashboardRequest = {
  name?: string;
  description?: string | null;
  visibility?: DashboardVisibility;
  refresh?: DashboardRefreshSettings;
  notify_on_failure?: boolean;
};

export async function publishDashboard(
  conversationId: string,
  fileId: string,
  body: PublishDashboardRequest,
): Promise<{ dashboard: PublishedDashboard; version: DashboardVersion }> {
  return request(
    `/api/chat/conversations/${encodeURIComponent(conversationId)}/files/${encodeURIComponent(fileId)}/publish-dashboard`,
    { method: "POST", body: JSON.stringify(body) },
  );
}

export async function listDashboards(options?: {
  includeArchived?: boolean;
}): Promise<{ dashboards: PublishedDashboard[] }> {
  const query = options?.includeArchived ? "?include_archived=true" : "";
  return request(`/api/dashboards${query}`);
}

export async function getDashboard(idOrSlug: string): Promise<DashboardDetail> {
  return request(`/api/dashboards/${encodeURIComponent(idOrSlug)}`);
}

export async function getDashboardBundle(
  dashboardId: string,
  versionId: string,
): Promise<DashboardBundle> {
  return request(
    `/api/dashboards/${encodeURIComponent(dashboardId)}/versions/${encodeURIComponent(versionId)}/bundle`,
  );
}

/** Fetch a version's stored dataset file with auth headers (for downloads). */
export async function fetchDashboardDataset(
  dashboardId: string,
  versionId: string,
  name: string,
): Promise<Blob> {
  const headers = await getAuthHeaders();
  const response = await fetch(
    `${GATEWAY_URL}/api/dashboards/${encodeURIComponent(dashboardId)}/versions/${encodeURIComponent(versionId)}/datasets/${encodeURIComponent(name)}`,
    { headers },
  );
  if (!response.ok) throw new Error(`Download failed (${response.status})`);
  return response.blob();
}

export async function updateDashboard(
  dashboardId: string,
  body: UpdateDashboardRequest,
): Promise<{ dashboard: PublishedDashboard }> {
  return request(`/api/dashboards/${encodeURIComponent(dashboardId)}`, {
    method: "PATCH",
    body: JSON.stringify(body),
  });
}

export async function refreshDashboardNow(
  dashboardId: string,
): Promise<{ refresh: DashboardRefresh }> {
  return request(`/api/dashboards/${encodeURIComponent(dashboardId)}/refresh`, {
    method: "POST",
  });
}

export async function listDashboardRefreshes(
  dashboardId: string,
  limit = 20,
): Promise<{ refreshes: DashboardRefresh[] }> {
  return request(
    `/api/dashboards/${encodeURIComponent(dashboardId)}/refreshes?limit=${limit}`,
  );
}

export async function restoreDashboardVersion(
  dashboardId: string,
  versionId: string,
): Promise<{ dashboard: PublishedDashboard; version: DashboardVersion }> {
  return request(
    `/api/dashboards/${encodeURIComponent(dashboardId)}/versions/${encodeURIComponent(versionId)}/restore`,
    { method: "POST" },
  );
}

export async function archiveDashboard(
  dashboardId: string,
): Promise<{ dashboard: PublishedDashboard }> {
  return request(`/api/dashboards/${encodeURIComponent(dashboardId)}/archive`, {
    method: "POST",
  });
}

export async function unarchiveDashboard(
  dashboardId: string,
): Promise<{ dashboard: PublishedDashboard }> {
  return request(`/api/dashboards/${encodeURIComponent(dashboardId)}/unarchive`, {
    method: "POST",
  });
}

export async function deleteDashboard(dashboardId: string): Promise<void> {
  await request(`/api/dashboards/${encodeURIComponent(dashboardId)}`, {
    method: "DELETE",
  });
}

export async function openDashboardEditChat(
  dashboardId: string,
): Promise<{ conversation_id: string }> {
  return request(`/api/dashboards/${encodeURIComponent(dashboardId)}/edit-chat`, {
    method: "POST",
  });
}

export const DASHBOARD_REFRESH_OPTIONS: { value: DashboardRefreshInterval | null; label: string }[] = [
  { value: null, label: "Off" },
  { value: 15, label: "Every 15 minutes" },
  { value: 60, label: "Every hour" },
  { value: 120, label: "Every 2 hours" },
  { value: 240, label: "Every 4 hours" },
  { value: 480, label: "Every 8 hours" },
  { value: 720, label: "Every 12 hours" },
  { value: 1440, label: "Daily" },
];
